### PR TITLE
Implement component retry properly

### PR DESCRIPTION
**Pull request by Others**
Now we retry if we 404 or if we can't find a worker. Only one retry per error condition is allowed.

The alternative here is just to call `update_workers` on a fast period. This has the advantage of removing a ton of code. However, in case of a read storm, it could cause requests to erroneously fail.

I also wonder if there is a locking strategy that would give us more faith in the state of the worker map. This code lets it change from under us -- which could cause unnecessary calls to `update_workers`. However, we probably don't want to hold the lock across a call to the worker (which could take an arbitrarily long time), so strategies in this direction complicate the code somewhat.

### CODE REVIEW

File: src/router.rs [hunk 1]
```diff
@@ -125,46 +126,101 @@ impl RequestForwarder {
...
-
-        let target_component = match target_component_option {
-            Some(v) => v,
-            None => {
-                let body = Body::from("Requested component not found\n");
-                warn!("Component to forward to not found: {:?}", component_name);
-                return Ok(Response::builder().status(404).body(body).unwrap());
-            }
-        };
+    fn find_appropriate_worker(&self, component_name: &str) -> Option<Arc<WorkerNode>> {
```
Comment (pcodes) [inline src/router.rs:129]: Glad you added this function, will be useful when we actually start doing educated decisions about which worker to send a request to

File: src/router.rs [hunk 2]
```diff
@@ -125,46 +126,101 @@ impl RequestForwarder {
...
+        component_name: &str,
+        worker_url: &str,
+    ) -> Result<Response<Body>, RouterError> {
+        let mut url = format!("{}/sl/{}/{}", worker_url, component_name, request.method);
 
         if !request.query.is_empty() {
             url = format!("{}?{}", url, request.query);
         }
 
+        // TODO: This blocks the executor, so we probably should do something smarter than just blocking
```
Comment (pcodes) [inline src/router.rs:147]: This won't scale to different connections at a time, will it? If not, we definitely need to find a better way

**Review by pcodes — APPROVED** [review]
Overall looks good, we discussed some future changes to be made that I'm gonna convert to issues. Otherwise, we resolved my questions IRL.

File: src/router.rs (hunk 2 above)
Comment (Others) [reply src/router.rs:147]: I think by default the executor has #threads = #cores. (And this only blocks one thread [I think]) So it isn't as bad as being single threaded. The right way to do this is probably to actually use tokio directly/make everything in an async context.

I want to try this in the worker first, so let's hold off fixing for now

**Comment by Others** [comment]
@pcodes Do you want me to squash and merge this?

**Comment by pcodes** [comment]
@Others yes please!

**Pull request merged**